File: 22. Generate Parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> list[str]:
        '''
        This uses a DFS helper method to solve the problem.
        '''
        if n == 0:
            return []
        left, right = n, n
        results = []
        self.helper(left, right, results, '')
        return results
        
    def helper(self, left, right, results, string):
        if right < left:
            return 
        if not left and not right:
            results.append(string)
            return
        if left:
            self.helper(left-1, right, results, string + '(')
        if right:
            self.helper(left, right-1, results, string + ')')

File: 22. Generate Parentheses.java
class Solution {
    private List<String> ans = new ArrayList<>();
    
    public List<String> generateParenthesis(int n) {
        genPar("", n, n);
        return this.ans;
    }
    
    private void genPar(String s, int left, int right) {
        if (left == 0 && right == 0) {
            this.ans.add(s);
            return;
        } 
        if (left > right || left < 0 || right < 0) {
            return;
        } else {
            genPar(s+'(', left-1, right);
            genPar(s+')', left, right-1);
        }
    }
}